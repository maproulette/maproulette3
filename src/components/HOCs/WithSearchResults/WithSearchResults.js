import React, { Component } from 'react'
import Fuse from 'fuse.js'
import _get from 'lodash/get'
import _isString from 'lodash/isString'
import _isArray from 'lodash/isArray'
import _isEmpty from 'lodash/isEmpty'
import _filter from 'lodash/filter'
import _find from 'lodash/find'
import _startsWith from 'lodash/startsWith'
import _omit from 'lodash/omit'
import WithSearchQuery from '../WithSearchQuery/WithSearchQuery'
import { parseQueryString } from '../../../services/Search/Search'

/**
 * WithSearchResults acts as a filter that applies the named search query to an
 * array of candidate items, presenting to the wrapped component only those
 * items that (fuzzily) match the query.
 *
 * @param {string} searchName - the name of the search/query to work with
 * @param {string} itemsProp - the name of the prop containing the array
 *        of items to search (e.g. 'challenges' or 'projects').
 * @param {string} outputProp - optional name of the prop to use when passing
 *        down the filtered results. By default it will be the same as
 *        itemsProp.
 */
const WithSearchResults = (WrappedComponent, searchName, itemsProp, outputProp) =>
  WithSearchQuery(
    _WithSearchResults(WrappedComponent, searchName, itemsProp, outputProp),
    searchName
  )

// Setup local fuzzy search configuration. See fusejs.io for details.
const searchOptions = {
  shouldSort: true, // sort results best to worst
  threshold: 0.4, // score: 0.0 is perfect, 1.0 terrible
  location: 0,
  distance: 100,
  maxPatternLength: 64, // max query length
  minMatchCharLength: 3,
  keys: ["name", "displayName", "blurb"], // fields to search
}

const _WithSearchResults = function(WrappedComponent, searchName, itemsProp, outputProp) {
  return class extends Component {
    /**
     * @private
     */
    itemsMatchingTags = (items, tagTokens) => {
      return _filter(items, (item) => {
        if (_isArray(item.tags)) {
          for (let tag of item.tags) {
            if (_find(tagTokens, (token) => _startsWith(tag, token))) {
              return true
            }
          }
        }

        return false
      })
    }

    render() {
      const query = _get(this.props, `searchQueries.${searchName}.searchQuery.query`, '')
      let items = this.props[itemsProp]
      let searchResults = this.props[itemsProp]

      if (_isString(query) && query.length > 0 &&
          _isArray(items) && items.length > 0) {
        const queryParts = parseQueryString(query)

        // If there are tags, filter the items up-front.
        if (queryParts.tagTokens.length > 0) {
          items = this.itemsMatchingTags(items, queryParts.tagTokens)
        }

        if (items.length > 0 && queryParts.query.length > 0) {
          const fuzzySearch = new Fuse(items, searchOptions)
          searchResults = fuzzySearch.search(queryParts.query)
        }
        else {
          searchResults = items
        }
      }

      if (_isEmpty(outputProp)) {
        outputProp = itemsProp
      }

      return <WrappedComponent {...{[outputProp]: searchResults}}
                               {..._omit(this.props, outputProp)} />
    }
  }
}

export default WithSearchResults
