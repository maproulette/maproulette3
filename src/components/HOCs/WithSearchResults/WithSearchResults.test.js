import { WithSearchResults } from './WithSearchResults'

let basicProps = null
let WrappedComponent = null

beforeEach(() => {
  basicProps = {
    searchCriteria: {
      query: ""
    },
    myItems: [
      {
        name: "first item",
        tags: ["foo"]
      },
      {
        name: "second item"
      },
      {
        name: "third item",
        tags: ["bar", "baz"]
      }
    ]
  }

  WrappedComponent = WithSearchResults(
    () => <div className="child" />,
    "mySearchName",
    "myItems"
  )
})

test("Search query is passed to wrapped component", () => {
  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  expect(wrapper.props().searchCriteria.query).toBe("")
  expect(wrapper.props().myItems.length).toBe(basicProps.myItems.length)

  expect(wrapper).toMatchSnapshot()
})

test("Search Results with tags are passed first in search results", () => {
  basicProps.searchCriteria.query = "#bar"

  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  expect(wrapper.props().myItems[0].name).toBe('third item')
  expect(wrapper.props().myItems[0].tags[0]).toBe('bar')
  expect(wrapper).toMatchSnapshot()
})

test("Search Results are passed back as the 'outputProp' if provided", () => {
  WrappedComponent = WithSearchResults(
    () => <div className="child" />,
    "mySearchName",
    "myItems",
    "myOutput"
  )

  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  expect(wrapper.props().myOutput.length).toBe(basicProps.myItems.length)
  expect(wrapper).toMatchSnapshot()
})
