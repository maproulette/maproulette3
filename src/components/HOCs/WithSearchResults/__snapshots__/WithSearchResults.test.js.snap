// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Results are passed back as the 'outputProp' if provided 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    myItems={
      Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ]
    }
    searchCriteria={
      Object {
        "query": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "myItems": Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ],
      "myOutput": Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ],
      "mySearchNameSearchActive": false,
      "searchCriteria": Object {
        "query": "",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "myItems": Array [
          Object {
            "name": "first item",
            "tags": Array [
              "foo",
            ],
          },
          Object {
            "name": "second item",
          },
          Object {
            "name": "third item",
            "tags": Array [
              "bar",
              "baz",
            ],
          },
        ],
        "myOutput": Array [
          Object {
            "name": "first item",
            "tags": Array [
              "foo",
            ],
          },
          Object {
            "name": "second item",
          },
          Object {
            "name": "third item",
            "tags": Array [
              "bar",
              "baz",
            ],
          },
        ],
        "mySearchNameSearchActive": false,
        "searchCriteria": Object {
          "query": "",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Search Results with tags are passed first in search results 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    myItems={
      Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ]
    }
    searchCriteria={
      Object {
        "query": "#bar",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "myItems": Array [
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ],
      "mySearchNameSearchActive": false,
      "searchCriteria": Object {
        "query": "#bar",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "myItems": Array [
          Object {
            "name": "third item",
            "tags": Array [
              "bar",
              "baz",
            ],
          },
        ],
        "mySearchNameSearchActive": false,
        "searchCriteria": Object {
          "query": "#bar",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Search query is passed to wrapped component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    myItems={
      Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ]
    }
    searchCriteria={
      Object {
        "query": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "myItems": Array [
        Object {
          "name": "first item",
          "tags": Array [
            "foo",
          ],
        },
        Object {
          "name": "second item",
        },
        Object {
          "name": "third item",
          "tags": Array [
            "bar",
            "baz",
          ],
        },
      ],
      "mySearchNameSearchActive": false,
      "searchCriteria": Object {
        "query": "",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "myItems": Array [
          Object {
            "name": "first item",
            "tags": Array [
              "foo",
            ],
          },
          Object {
            "name": "second item",
          },
          Object {
            "name": "third item",
            "tags": Array [
              "bar",
              "baz",
            ],
          },
        ],
        "mySearchNameSearchActive": false,
        "searchCriteria": Object {
          "query": "",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
